sdk/src/main/java/io/horizen/sc2sc/Sc2ScUtils.scala

```diff
@@ -50,69 +57,86 @@ trait Sc2ScUtils[
-          )
-          val topCertMerklePath = MerklePath.deserialize(topCertInfos.merklePath)
-          val topCertscCommitmentRoot = FieldElement.deserialize(topCertInfos.scCommitmentRoot)
-          val nextCertMerklePath = MerklePath.deserialize(nextTopCertInfos.merklePath)
-          val nextCertscCommitmentRoot = FieldElement.deserialize(nextTopCertInfos.scCommitmentRoot)
-
-
-          //compute proof
-          val proof :Array[Byte]= sc2scCircuitFunctions.createRedeemProof(
-             messageHash,
-             messageMerklePath,
-             CommonCircuit.createWithdrawalCertificate(topCertInfos.certificate, params.sidechainCreationVersion),
-             topCertMerklePath,
-             topCertscCommitmentRoot,
-             CommonCircuit.createWithdrawalCertificate(nextTopCertInfos.certificate, params.sidechainCreationVersion),
-             nextCertMerklePath,
-             nextCertscCommitmentRoot,
-          )
-
-          //build and return final message
-          val retMessage = new CrossChainRedeemMessageImpl(
-             sourceMessage,
-             CryptoLibProvider.commonCircuitFunctions.getCertDataHash(topCertInfos.certificate, params.sidechainCreationVersion),
-             CryptoLibProvider.commonCircuitFunctions.getCertDataHash(nextTopCertInfos.certificate, params.sidechainCreationVersion),
-             topCertInfos.scCommitmentRoot,
-             nextTopCertInfos.scCommitmentRoot,
-             proof
-          )
-          Success(retMessage)
+          Using.resource(
+            sc2scCircuitFunctions.initMerkleTree()
+          ) { tree => {
+            val messages = state.getCrossChainMessages(messagePostedEpoch).asJava
+            val msgLeafIndex = sc2scCircuitFunctions.insertMessagesInMerkleTreeWithIndex(tree, messages, sourceMessage)
+            val messageMerklePath: MerklePath = sc2scCircuitFunctions.getCrossChainMessageMerklePath(tree, msgLeafIndex)
+            val topCertInfos = getTopCertInfoByEpoch(messagePostedEpoch, state, history, calculateMerklePath = true).getOrElse(
+              throw new Sc2ScException("Unable to retrieve certificate associated with this message epoch")
+            )
+            val nextTopCertInfos = getTopCertInfoByEpoch(messagePostedEpoch + 1, state, history, calculateMerklePath = true).getOrElse(
+              throw new Sc2ScException("Unable to retrieve certificate associated with this message epoch+1")
+            )
+            val topCertscCommitmentRoot = topCertInfos.scCommitmentRoot
+            val nextCertscCommitmentRoot = nextTopCertInfos.scCommitmentRoot
+
+            val currWithdrawalCertificate = CommonCircuit.createWithdrawalCertificate(topCertInfos.certificate, params.sidechainCreationVersion)
+            val nextWithdrawalCertificate = CommonCircuit.createWithdrawalCertificate(nextTopCertInfos.certificate, params.sidechainCreationVersion)
+
+            val currentCt = CommitmentTree.init()
```

Comment:
As we discuess this path should be the path from the certificate to the scTxCommitmentTreeRoot